### PR TITLE
Unidoc: Changed processfile in UniAll to support ulsp

### PR DESCRIPTION
**Pull request by DuyMach**
Add new method in UniAll to grab line directly from raw buff inside UniFile.
Changed UniFile to be able to update (Support for in memory processing)

### CODE REVIEW

**Comment by DuyMach** [comment]
fixed issue from PR, awaiting review

File: uni/unidoc/UniAll.icn [hunk 1]
```diff
@@ -151,22 +151,30 @@ class UniAll : Object (files, packages,
...
         if stat(fName||"/") then {   # Handle source files in directory
            every processFile(hasSuffix(!open(fName), ".icn"))
            return
            }
 
-        fName := delSuffix(fName,".icn")||".icn"
-        uniFile := UniFile(fName) |
+           fName := delSuffix(fName, ".icn") || ".icn"
+
+           uniFile := UniFile(fName) |
```
Comment (Jafaral) [inline uni/unidoc/UniAll.icn:165]: wrong indentation

File: uni/unidoc/UniAll.icn [hunk 2]
```diff
@@ -151,22 +151,30 @@ class UniAll : Object (files, packages,
...
                    UniFile(fName, genDirs(\sourcePath)) | {
                    insert(badFiles, fName)
                    fail
-                   }
-        if f := addFile(fName) then {
+                   } 
+
+        if \contents then {
+            if /uniFile := uniFilez[fName] then {
```
Comment (Jafaral) [inline uni/unidoc/UniAll.icn:172]: why do you check for null in `/uniFile`?  do you only want to run this if uniFile is null ?

File: uni/unidoc/UniAll.icn [hunk 3]
```diff
@@ -151,22 +151,30 @@ class UniAll : Object (files, packages,
...
+                uniFile.updateContent(contents) | continue
+                classifyLines(uniFile)
+                uniFilez[fName] := uniFile
+            }
+        } else if f := addFile(fName) then {
             uniFile.setSaveSrc(saveSrcFlag)
```
Comment (Jafaral) [inline uni/unidoc/UniAll.icn:178]: did you mean to do `uniFile := addFile(fName)` here? `f` isn't used.

File: uni/unidoc/UniAll.icn [hunk 4]
```diff
@@ -199,9 +208,17 @@ class UniAll : Object (files, packages,
                 cleanSets()
                 }
 
-            }
-            
+            }
```
Comment (Jafaral) [inline uni/unidoc/UniAll.icn:211]: remove whitespace at the end of the line.

File: uni/unidoc/UniAll.icn [hunk 5]
```diff
@@ -199,9 +208,17 @@ class UniAll : Object (files, packages,
...
-            
+            }  
+    end
 
+    #<p>
+    #    Given the name of a file and line number, retrieve the specific
+    #    line directly from the raw buffer inside UniFile.
+    #</p>
+    method getUniFileLine(fName, lineNum)
+        if /fName | (*fName = 0) then fail
```
Comment (Jafaral) [inline uni/unidoc/UniAll.icn:219]: you can shorten this as  if (*\fName) = 0 then fail

File: uni/unidoc/UniAll.icn [hunk 6]
```diff
@@ -1169,4 +1188,4 @@ procedure mkListStr(sequence)
             }
         }
     return (s[1:-2] | "")
-end
+end
```
Comment (Jafaral) [inline uni/unidoc/UniAll.icn:1191]: add an empty line at the end of the file.

File: uni/unidoc/UniFile.icn [hunk 7]
```diff
@@ -24,11 +24,11 @@ import lang
 #   a format suitable for output (e.g. an HTML document), but only if the
 #   call <b>setSaveSrc("yes")</b> is made on this class.
 # </p>
-class UniFile : Object (iKeySet, uKeySet, oldTokens, buffer, line, f, fName,
-                       insideFlag, debug, saveSrc, cBuf)
+class UniFile : Object (iKeySet, uKeySet, oldTokens, buffer, line, fName,
+                       insideFlag, debug, saveSrc, cBuf, isOpen, lineNum)
```
Comment (Jafaral) [inline uni/unidoc/UniFile.icn:28]: isOpen is not used.

File: uni/unidoc/UniFile.icn [hunk 8]
```diff
@@ -305,14 +304,6 @@ class UniFile : Object (iKeySet, uKeySet, oldTokens, buffer, line, f, fName,
         return buffer := s
     end
 
-    # <p>
-    # Close this file.
-    # </p>
-    method close()
-        cBuf := &null
-        ::close(f)
-    end
```
Comment (Jafaral) [inline uni/unidoc/UniFile.icn:314]: why are  you dropping this function.

File: uni/unidoc/UniFile.icn [hunk 9]
```diff
@@ -327,18 +318,41 @@ class UniFile : Object (iKeySet, uKeySet, oldTokens, buffer, line, f, fName,
...
+    method getFromFile(f)
+        cBuf := CodeBuf()
+        while line := read(f) do cBuf.add(line)
+        close(f)
+    end
+
     # <b>Fails if cannot open <tt>fName</tt> for reading.</b>
     initially (fileName, path)
-        /path := ""
+        /path:= ""
```
Comment (Jafaral) [inline uni/unidoc/UniFile.icn:345]: re-add the space before the colon.

File: uni/unidoc/UniFile.icn [hunk 10]
```diff
@@ -377,6 +392,16 @@ class CodeBuf : Object (buffer)
...
+
+    method Get(i)
+        return rawBuf[i] | fail
+    end
+
     initially ()
         buffer    := []
-end
+        rawBuf    := []
+end
```
Comment (Jafaral) [inline uni/unidoc/UniFile.icn:407]: add a new line at the end of the file.

**Review by Jafaral — CHANGES_REQUESTED** [review]

**Review by Jafaral — APPROVED** [review]

**Pull request merged**